DS-459 Add Dialog component

danielamorse opened the pull request:
## Jira

https://pegadigitalit.atlassian.net/browse/DS-459

## Summary

Add new experimental Dialog component. It is _not_ officially ready to be used.

## Details

Adds new Dialog component. It is not a web component. It includes support for the same props as Bolt Modal except for scroll behavior (dialog container scrolls by default).

This uses the new `<dialog>` element and its associated browser API. It is supported by all modern browsers, but for Safari it is only supported in 15.4 and up. We are merging this work now but marking in the docs that it is not officially released until adoption of the latest versions of Safari improves.

There is no JS documentation in PL yet because you can not easily call the Dialog methods from outside of Bolt. This is because the JS in Bolt is loaded asynchronously, which means you need to also add some event listener to know if/when the component has been loaded, and it quickly gets complicated. I'm going to look into an alternate approach using [Drupal libraries and behaviors](https://pegadigitalit.atlassian.net/browse/DS-783).

One other known issue is the close button positioning, see https://pegadigitalit.atlassian.net/browse/DS-782.

## How to test

- Review code and docs

## Release notes

### Dialog
- Dialog component is added. It is _not_ ready for sites to use until Safari 15.4 adoption improves.

colbytcook commented:
@danielamorse I am not seeing the `content` description being displayed for the "Dialog Header", "Dialog Body", or "Dialog Footer" schema section in Patterrnlab. Not sure if we care about this or not but I figured I would mention it.

colbytcook commented on docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-05-dialog.twig:
```diff
@@ -0,0 +1,79 @@
+  description: description,
+  notes: notes,
+  demo: demo,
+  twig_markup: twig_markup,
```
@danielamorse I am not seeing this markup example show in Patternlab

colbytcook commented:
@danielamorse I am seeing a mobile video display error, see the attached screenshot.
<img width="1662" alt="mobile-video-error" src="https://user-images.githubusercontent.com/1932313/169852088-f3431f72-42c2-4fb4-a857-a8a8c3f44051.png">

danielamorse commented:
> @danielamorse I am seeing a mobile video display error, see the attached screenshot. <img alt="mobile-video-error" width="1662" src="https://user-images.githubusercontent.com/1932313/169852088-f3431f72-42c2-4fb4-a857-a8a8c3f44051.png">

@mikemai2awesome the dialog has no inline size until `small` breakpoint. Should it be set to 100% [here](https://github.com/boltdesignsystem/bolt/pull/2492/files#diff-8544735a8ef8b261f6071d7b6daa7d2e789cf4ea28199febca2d5b56c17a405eR105-R111)?

danielamorse commented:
> @danielamorse the close button on an image modal is hard to see, see screenshot. <img alt="image-close-bug" width="1661" src="https://user-images.githubusercontent.com/1932313/169852870-37b20458-6db3-4e3f-b0b9-e53ad2f73ec2.png">

@colbytcook this will be addressed in https://pegadigitalit.atlassian.net/browse/DS-782.

colbytcook commented on docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-20-dialog-persistent.twig:
```diff
@@ -0,0 +1,82 @@
+    id: 'dialog-persistent-1'
+  },
+} only %}
+{% endset %}
```
Is this an extra {% endset %}?

colbytcook replied:
disregard, this is the closing demo set tag, the indents messed me up

colbytcook commented:
I am seeing a bug on the "/pattern-lab/?p=tests-dialog-in-carousel" page. If you click to the second slide and open tthe carousel the modal will be shown over the first slide and not the second (if  you click back the first slide will see a cropped version of the modal)

colbytcook commented on packages/components/bolt-dialog/src/dialog-footer.twig:
```diff
@@ -0,0 +1,17 @@
+  'c-bolt-dialog__footer',
+] %}
+
+<div {{ attributes.addClass(classes) }}>
```
@danielamorse Should this be wrapped in a <footer> like the header is wrapped in a <header>?

danielamorse commented:
@colbytcook
**Mobile video display**
Nice catch. Fixed, revealed another issue we can handle separately (https://pegadigitalit.atlassian.net/browse/DS-784).

**Carousel Modal**
The second slide is the old Modal. It is expected to break. I updated the button text so that's more clear.

**Footer**
I'm not sure. The docs lead me to think "it's not that kind of footer". Then again, we _are_ using `<header>` here. @mikemai2awesome?

colbytcook reviewed (approved):
@danielamorse with the two follow tickets (video on mobile and image close button visibility) every looks good to me. Approved

mikemai2awesome reviewed (approved):
Ok to merge. Follow up tickets will take care of remaining issues.
